### ISSUE TITLE
Translate custom blocks

### ISSUE THREAD

**Issue by olafgrabienski** [backdrop/backdrop-issues#2956]
Like nodes, custom blocks should be directly translatable by Backdrop core.

The current workaround to translate custom blocks is: create a block for each language and add language related visibility conditions when placing these blocks in the Layout UI. This isn't straightforward, and it doesn't work well at all for websites with many languages or many custom blocks.

We discussed various approaches to the translation of custom blocks [in the Backdrop4Good queue](https://github.com/Backdrop4Good/backdrop4good.org/issues/39), for a summary see https://github.com/Backdrop4Good/backdrop4good.org/issues/39#issuecomment-353875203:

(1) Collapsible block groups
- didn't find favor

(2) [[CMI] Add config translation](https://github.com/backdrop/backdrop-issues/issues/704)
- promising for configuration translation in general
- probably not the right thing for *custom* blocks

(3) Use nodes for blocks
- not about custom blocks but a workaround for advanced site builders
- recommended by many, has benefits like fieldability
- doesn't however work with translation yet (I'll file a respective issue later)

(4) Make blocks directly translatable
- from an editor role perspective the most straightforward approach

In the latest weekly dev meetings (see e.g. https://youtu.be/mOkEj1ija0M?t=5m36s) we came to the conclusion to concentrate on the last mentioned approach: Make blocks directly translatable.

We don't have a PR so far but a working prototype by @bjoern-st. It's a rewrite of the Block module ( https://github.com/bjoern-st/block, first discussion here: https://github.com/Backdrop4Good/backdrop4good.org/issues/39#issuecomment-352896233).

**Comment by olafgrabienski** [comment]
@bjoern-st Are you still up to take your rewrite of the Block module and make a PR against core of it? I think that would be a great start for this issue even if you won't have the time to work on it continuously.

**Comment by olafgrabienski** [comment]
Linking to a closed duplicate so that we get a reference link to this issue over there: https://github.com/backdrop/backdrop-issues/issues/1576

**Comment by bjoern-st** [comment]
@olafgrabienski Yes. I have an idea how to to solve the things I felt bad with. 
I have some time today, so maybe I'll have some results this evening. 
I'll leave a comment here.

**Comment by bjoern-st** [comment]
@olafgrabienski 
Work is almost done. By now the only thing on my ToDo-list is, that the block is loaded in the right language on the Layout-Page for managing blocks. 
The rest is done and can be tested. 

Cheers Bjoern

**Comment by olafgrabienski** [comment]
@bjoern-st I'm available for testing now! Have you already made a PR against core?

**Comment by bjoern-st** [comment]
@olafgrabienski No. I would like to fix the language problem on the Layout page before the PR. 
Shouldn't be a big deal - maybe I can fix it this evening.

**Comment by olafgrabienski** [comment]
@bjoern-st Ok, so I'll test it then. And thanks for working on it!

**Comment by bjoern-st** [comment]
@olafgrabienski I've tested and rewritten my code an now everything works like expected and works fine with the layout module, too. 
So I made a PR 
https://github.com/backdrop/backdrop/pull/2091

It seems that there are some problems with the php7 test - I'll have a look at it tomorrow. 

Cheers Bjoern

**Comment by olafgrabienski** [comment]
@bjoern-st Nice, I'm looking forward to test the PR tonight or tomorrow!

**Comment by olafgrabienski** [comment]
I've had a look at the sandbox site of the PR, the functionality looks good:

- I was able to create a custom block in English and to translate it to French and German.
- It was also possible to place the block in the default layout to make sure that the specific language blocks showed up on respective pages.

The translation process was quite straightforward, but I think there's room to improve the user interface. I'd suggest to adapt the translate blocks interface to the translation of nodes, cf. for instance:

- the behaviour of the language selector on the edit form (always available for nodes vs. only while creating for blocks)
- the name of the language selector when no language is selected (` - None - ` vs. `Language neutral`)
- column number and names on translation overview pages (see comparison below)

Translate node overview page:
![screen-backdrop-pr-2091-translate-node-overview](https://user-images.githubusercontent.com/14188307/35783539-4d17dc3a-0a09-11e8-911c-658208967d55.png)


Translate custom block overview page:
![screen-backdrop-pr-2091-translate-block-overview](https://user-images.githubusercontent.com/14188307/35783538-4cfaa46c-0a09-11e8-84da-3835e2b12f2e.png)

**Comment by olafgrabienski** [comment]
Regarding the placement of a translated block in a layout: As mentioned in the comment above, I was able to place the block in a layout, and the specific language blocks showed up on respective language specific pages.

While editing the custom block later from the layout page, the block content showed up in the language of the admin interface. So, when I was on a French admin page, the block content was displayed in French as well, very nice!

I noted a common use case which doesn't work very smoothly though: to display a block on a particular page, I use to add a visibility condition like `Current path is one of "node/3"`. With a translated block I have to mention the paths of the translated nodes as well, e.g.: `Current path is one of "node/3, node/4 , node/5"`. In case of many translations, an option to only mention the 'original' node (which then had to follow the translation system) would be helpful. Probably a separate issue not to be solved by the current PR.

**Comment by herbdool** [comment]
@olafgrabienski That last issue you mention is similar to how it worked in Drupal 7. It would be a complicated solution at any rate. It needs to be able to know if you want the block to show up for all translated nodes in the set or for only one of the languages. It can't assume the former.

**Comment by herbdool** [comment]
@bjoern-st @olafgrabienski I noticed that unless the custom block is set to reusable there is no option to translate. For instance, people can add custom blocks from the layout. One can create a block from the layout and not know that it can be translated. 

After creating the block from the layout, one needs to check the checkbox to make it reusable. After checking that they to the custom block page and translate the block from there.

Maybe even a message telling people they need to make it reusable in order to translate and go to the custom block page can be helpful. Only if languages are enabled.

**Comment by bjoern-st** [comment]
@herbdool Hello Herb

Thank you for your feedback. I didn't notice that you can create custom blocks in the layout form. 
But maybe this is a general problem because if you don't mark that block as reusable it doesn't even show up at the custom block list. 
So a notice would really be a great idea. 

I'll integrate that in my PR.

Cheers Bjoern

**Comment by bjoern-st** [comment]
@olafgrabienski OK, you are right. The form for node translation is more intuitive and straightforward. 
I'll change that and will make a new PR.

Cheers Bjoern

**Comment by herbdool** [comment]
@bjoern-st great work and thanks!

**Comment by olafgrabienski** [comment]
@herbdool and @bjoern-st Thanks for your comments! I'm on a journey for a couple of weeks and have only limited time, will follow the discussion more silently in this period.

**Comment by bjoern-st** [comment]
@olafgrabienski @herbdool 
I've integrated the changes and made an new PR. See
https://github.com/backdrop/backdrop/pull/2091

Cheers Bjoern

**Comment by olafgrabienski** [comment]
@bjoern-st Thanks for updating the PR! Will try to give some feedback regarding the UI in comparison to node translation: it became already more similar, there are however still some differences, e.g. the language selector on the edit form which is still not available while *editing* blocks. Actually, there are more UI differences, I guess I didn't mention all of them in https://github.com/backdrop/backdrop-issues/issues/2956#issuecomment-362948890:

- The title of the Translate block page is "Configure block", I'd suggest something like "Translations of [BLOCK NAME]".
- The title of the translate tab is "Translate blocks" but should be singular ("Translate block"), or even just "Translate".
- While in the "Status" column of a node translation we have "Not translated" or "Published", and the latter applies also to the source node, with the current PR the source block has the status "Not Translated". There's also an "Add translation" link for the source block, an "Edit" link makes more sense in this case.

Probably not directly related: The title of the source block isn't displayed on the Translate block page, I see just a "N/A" in the column.

**Comment by bjoern-st** [comment]
@olafgrabienski @herbdool
I made the changes and updated the PR. Now all the tests are passed successfully.
http://2091.backdrop.backdrop.qa.backdropcms.org/

**Comment by herbdool** [comment]
@bjoern-st @olafgrabienski Looks good. I think it addresses Olaf's suggestions. Thanks!

**Comment by olafgrabienski** [comment]
@bjoern-st Sounds good! I‘m on the road these days, lacking internet access. will have a look by monday or tuesday.

**Comment by olafgrabienski** [comment]
Here I am again. Thanks @bjoern-st for the updated PR and thanks @herbdool for the review! I've had a look at the sandbox site, and the custom block translation looks very good now in my opinion! I've found two minor issues in help texts though:

- Adding a block via the Custom blocks page, on the block add form in the phrase `Please select the default language for this custum block.` the "cust**u**m" (with an `u`) should be "cust**o**m" (with an `o`).
- Adding a block via a Layout, there is now the new help text with a hint to translation as discussed above: `NOTE: You have module "language" enabled. If you want to make this block translatable, this block must be marked as reusable.` It's an important notice, although in my opinion the "NOTE" in capitals and the first part about the enabled language module are not necessary. Here's a shorter suggestion (within context):

    `[ ] Make this block reusable`
    `If enabled, this block will be made reusable across different layouts and be listed in on the Custom blocks page.`
    `To make this block translatable, this block must be made reusable as well.`

**Comment by olafgrabienski** [comment]
Note to myself: make a suggestion about the option to mark a translatable block to follow the translation system.

Related issue for existing content blocks: https://github.com/backdrop/backdrop-issues/issues/2957

**Comment by bjoern-st** [comment]
@olafgrabienski 

I'll make the text adaptations asap and will report here when its done.

**Comment by bjoern-st** [comment]
@olafgrabienski 

Done. Updated the PR.

**Comment by herbdool** [comment]
I read the updated PR. Looks good.

**Comment by olafgrabienski** [comment]
Had a look at the sandbox site, looks also good! (Not all tests are passing but may be unrelated to the last changes.)

Thanks again for the great work! Looking forward to the integration of block translation in 1.10!

**Comment by bjoern-st** [comment]
@olafgrabienski 
Since the tests have been successful before and I only made text changes, it doesn't make sense that the tests were not successful. 
I also can't click on the "Detail" links to see the errors - so it might be an error in the test system.

**Comment by bjoern-st** [comment]
@olafgrabienski 
There was indeed a problem in the test area. Now all tests passed successfully again.

**Comment by quicksketch** [comment]
I'll take a look at this ASAP. Thanks @bjoern-st and @olafgrabienski for your work here!

**Comment by quicksketch** [comment]
I finally posted comments to the PR: https://github.com/backdrop/backdrop/pull/2091#pullrequestreview-110667376

The PR is noticeably missing tests as well, for a feature expansion like this we need to add test coverage. Scenario should at least include:

- Enabling language module (via API)
- Adding a second language (via API).
- Adding a block to a layout and making it reusable (via API).
- Translating the block via the new Block translation UI.
- Ensuring the translated block version shows up on the language-variant of the page.

**Comment by bjoern-st** [comment]
@quicksketch 
Thanks for your review.
I'll take care of your requested changes as soon as possible. 
I never implemented test for Drupal/Backdrop. So that point might take a while.

**Comment by herbdool** [comment]
@bjoern-st  I can help write tests if you're getting bogged down. I think I recall you mentioning somewhere that you were close (can't find the comment now)

**Comment by bjoern-st** [comment]
@herbdool 
Hey, thank you very much. Yes, you can help
The actual work is here https://github.com/bjoern-st/block

It passes all the tests but now i need to change the "default language" in the testing area and will have to check, if the right translation is shown ..

Cheers Bjoern

**Comment by bjoern-st** [comment]
Hey @herbdool 
I think a got it. As @jenlampton suggested I had a look at the locale module an found the way to set the default language. 
Now the tests just work fine.

Cheers Bjoern

**Comment by bjoern-st** [comment]
Things happend:
 - I made all the changes requested by @quicksketch here https://github.com/backdrop/backdrop/pull/2091
 - added tests 

Additional changes:
 - the block translation is now triggered by the locale module. 

Hope, that everything is in a good state now.

Bjoern

**Comment by olafgrabienski** [comment]
@bjoern-st I'd like to have another look at the sandbox site with the last changes but I saw that the PR has been closed, so there is no sandbox at the moment.

**Comment by bjoern-st** [comment]
@olafgrabienski 
The new PR with the sandbox is here https://github.com/backdrop/backdrop/pull/2153

**Comment by olafgrabienski** [comment]
Thanks @bjoern-st , I've found it and had a look!

Strangely, I found a small issue which I thought had already been fixed, cf. https://github.com/backdrop/backdrop-issues/issues/2956#issuecomment-368921115:

- Adding a block via the Custom blocks page, on the block add form in the phrase `Please select the default language for this custum block.` the "cust**u**m" (with an `u`) should be "cust**o**m" (with an `o`).

Regarding the questions brought up by @herbdool in the PR comments, I think the PR needs some work. Will make a comment over there.

**Comment by quicksketch** [comment]
Thanks @bjoern-st! Also I made you a collaborator so you can now add/remove tags, sorry that is still a manual process right now to add people. I'll make another review of it.

**Comment by quicksketch** [comment]
I've been working on rerolling this to fix my own feedback as well as that of @herbdool.

I ran into one UX question: Is it worth having a new, separate permission for "translate blocks"? As a translator, this would allow you to edit *any* translation (including the source). Which means you could just as well delete or add new blocks by wiping out the contents or filling in translations with junk.

**Comment by herbdool** [comment]
@quicksketch good question. The permission could always be added later. I was wondering how the "translate content" permission works; does it allow one to delete or edit the source content?

**Comment by quicksketch** [comment]
> does it allow one to delete or edit the source content?

It does allow you to edit the source content. But it does not prevent you from editing it and then wiping out its contents. Right now as a translator with "translate blocks" but not "administer blocks", you can't actually view the block listing at admin/structure/block either, which is another problem.

**Comment by herbdool** [comment]
Is the idea that "translate blocks" permission would be required in addition to "administer blocks"? Seems like administer would include translation too. Maybe "administer blocks" needs to be renamed too; since it's just adding and editing blocks.

**Comment by quicksketch** [comment]
> Is the idea that "translate blocks" permission would be required in addition to "administer blocks"? 

That seems to be the current situation. It's not useful at this point to be able to translate-only because you can't see the list of all blocks in the first place. And "administer blocks" does not automatically encompass "translate blocks", so you could in theory be able to create/edit/delete (and even set the source language), but you couldn't actually do the translation. Likewise, with the current approach of setting a source language before translation becomes available, a user with "administer blocks" would must set the default language before a different user with "translate blocks" could translate it.

These two permissions seem so intertwined at this point I think it would be better to nix "translate blocks" and just make it automatically available to those with "administer blocks".

**Comment by olafgrabienski** [comment]
I'm very happy about the progress of the last days, looks great! I've have had another look at the sandbox site today, added test blocks via the Custom block page and via a Layout page, translated them and saw that they show up on respective pages of different languages. All the things without any problem.

The UI feels more consistent now, and I agree with the removal of the "Translate blocks" permission for the time being.

Thanks to all of you!

**Comment by bjoern-st** [comment]
I included the requested changes an made a new PR.

Included changes:
 - added admin description field as in issue #2722 
 - hide translation option if no language is set 
 - added access callback to hide translation tab if no language is set 
 - remove block without language from translation list
 - fixed several typos and coding standards

**Comment by quicksketch** [comment]
Thanks all! @bjoern-st we may have duplicated efforts. I checked your most recent changes in the PR at https://github.com/backdrop/backdrop/pull/2153. They mostly cover the same changes I made in https://github.com/backdrop/backdrop/pull/2163. I'm sorry you may have spent time doing the same thing over again. :frowning_face: 

That said, it seems that your new updates are completely inline with the changes I made. So it appears as though we are all on the same page as far as what we want and how to accomplish it. With both @olafgrabienski and @herbdool having reviewed https://github.com/backdrop/backdrop/pull/2163, I would like to merge that PR. It includes some block loading/saving adjustments that differ, but the functional result is the same.

**Comment by quicksketch** [comment]
In the interest of make sure this gets into 1.10, I've merged https://github.com/backdrop/backdrop/pull/2163. Thanks so, so much @bjoern-st for your amazing work here! Thanks to @olafgrabienski and @herbdool for the fantastic collaboration and code reviewing.
